code-analysis/crates/codegen/parser/src/chumsky/boilerplate.rs

```diff
@@ -17,41 +17,37 @@ pub fn lex_head() -> TokenStream {
 
         #[derive(Clone, Debug, PartialEq, Eq, Serialize)]
         pub enum Node {
-            None,
             Chars(Range<usize>),
             Choice(usize, Rc<Node>),
             Sequence(Vec<Rc<Node>>),
             Named(kinds::Token, Rc<Node>),
         }
 
         impl Node {
-            pub fn none() -> Rc<Self> {
-                Rc::new(Self::None)
+            pub fn chars(range: Range<usize>) -> Option<Rc<Self>> {
+                Some(Rc::new(Self::Chars(range)))
             }
 
-            pub fn chars(range: Range<usize>) -> Rc<Self> {
+            pub fn chars_unwrapped(range: Range<usize>) -> Rc<Self> {
                 Rc::new(Self::Chars(range))
             }
 
-            pub fn sequence(elements: Vec<Rc<Self>>) -> Rc<Self> {
-                Rc::new(if elements.is_empty() {
-                    Self::None
-                } else {
-                    Self::Sequence(elements)
-                })
+            pub fn sequence(elements: Vec<Option<Rc<Self>>>) -> Option<Rc<Self>> {
+                let elements: Vec<_> = elements.into_iter().filter_map(|e| e).collect();
+                if elements.is_empty() { None } else { Some(Rc::new(Self::Sequence(elements))) }
             }
 
-            pub fn choice(number: usize, element: Rc<Self>) -> Rc<Self> {
-                Rc::new(Self::Choice(number, element))
+            pub fn choice(number: usize, element: Option<Rc<Self>>) -> Option<Rc<Self>> {
```

Comment:
Question: is `number` the index of the choice in the grammar? should it be `index` instead?

Reply:
Good catch